Message can be discarded during handshake due to delays on the message bus and wrong connection state

Situation: during the end-to-end tests evaluation with P4OfSwitch I noticed some tests were failing with errors like this:

```
tests/test_e2e_50_maintenance.py::TestE2EMaintenance::test_090_end_non_existent_running_mw_on_switch_should_fail: 2026-04-25,03:02:41.687318 - 2026-04-25,03:03:29.795118
self = <tests.test_e2e_50_maintenance.TestE2EMaintenance object at 0x7f51901ee210>

    def test_090_end_non_existent_running_mw_on_switch_should_fail(self):
        """
        404 response calling
            /api/kytos/maintenance/v1/{mw_id}/end on PATCH
        """
>       self.net.restart_kytos_clean()

tests/test_e2e_50_maintenance.py:957:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/helpers.py:483: in restart_kytos_clean
    self.wait_kytos_links()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.helpers.NetworkTest object at 0x7f5190149410>

    def wait_kytos_links(self):
        wait_count = 0
        last_error = ""
        topo_links = []
        for link in self.net.links:
            if link.intf1.node in self.net.switches and link.intf2.node in self.net.switches:
                link_id = self.create_link_id(link)
                if link_id:
                    topo_links.append(link_id)
        while wait_count < 30:
            try:
                response = requests.get("http://127.0.0.1:8181/api/kytos/topology/v3/links/", timeout=3)
                links = response.json()["links"]
                assert len(topo_links) == len(links), f"{topo_links=} {links=}"
                assert all(link_id in links for link_id in topo_links), f"{topo_links=} {links=}"
                break
            except Exception as exc:
                last_error = str(exc)
            time.sleep(1)
            wait_count += 1
        else:
            msg = f"Timeout waiting for links. Last error: {last_error}"
>           raise Exception(msg)
E           Exception: Timeout waiting for links. Last error: topo_links=['78282c4d5b579265f04ebadc4405ca1b49628eb1d684bb45e5d0607fa8b713d0', '4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30', 'c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07'] links={'4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30': {'id': '4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30', 'endpoint_a': {'id': '00:00:00:00:00:00:00:03:2', 'name': '2/0', 'port_number': 2, 'mac': '00:00:00:00:00:00', 'switch': '00:00:00:00:00:00:00:03', 'type': 'interface', 'nni': True, 'uni': False, 'speed': 12500000000.0, 'metadata': {}, 'lldp': True, 'active': True, 'enabled': True, 'status': 'UP', 'status_reason': [], 'link': '4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30'}, 'endpoint_b': {'id': '00:00:00:00:00:00:00:02:3', 'name': '3/0', 'port_number': 3, 'mac': '00:00:00:00:00:00', 'switch': '00:00:00:00:00:00:00:02', 'type': 'interface', 'nni': True, 'uni': False, 'speed': 12500000000.0, 'metadata': {}, 'lldp': True, 'active': True, 'enabled': True, 'status': 'UP', 'status_reason': [], 'link': '4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30'}, 'metadata': {}, 'active': True, 'enabled': True, 'status': 'UP', 'status_reason': []}}
```

With the RERUN strategy, eventually those tests will pass. However it called my attention that this was happening with a non-negligible frequency, then I decided to dig deeper to understand what was happening. After enabling a lot of logs to enable troubleshooting I got one case that seems to be the root cause. During this particular exec, I noticed that two switches connected out of three, so one switch didnt finish the connection and then the links didnt get fully discovered:

```
# egrep "^2026-04-25T11:31:[1-5]" /var/log/syslog| egrep "Starting Kytos - K|Stopping Kytos|HANDSHAKE"
2026-04-25T11:31:12.566943+00:00 ceccb6b6faa2 kytos.core.controller:INFO controller:386:  Starting Kytos - Kytos Controller
2026-04-25T11:31:18.529391+00:00 ceccb6b6faa2 kytos.napps.kytos/of_core:INFO main:171:  Connection ('172.18.0.2', 43526), Switch 00:00:00:00:00:00:00:01: OPENFLOW HANDSHAKE COMPLETE
2026-04-25T11:31:20.857231+00:00 ceccb6b6faa2 kytos.napps.kytos/of_core:INFO main:171:  Connection ('172.18.0.3', 54862), Switch 00:00:00:00:00:00:00:02: OPENFLOW HANDSHAKE COMPLETE
2026-04-25T11:31:54.715835+00:00 ceccb6b6faa2 kytos.core.controller:INFO kytosd:115:  Stopping Kytos controller...
2026-04-25T11:31:54.715934+00:00 ceccb6b6faa2 kytos.core.controller:INFO controller:515:  Stopping Kytos
```

With the new logs I've enabled, I checked for all messages from the IP that was missing:

```
root@ceccb6b6faa2:/kytos-end-to-end-tests# egrep "^2026-04-25T11:31:[1-5]" /var/log/syslog| grep 172.18.0.4
2026-04-25T11:31:19.834912+00:00 ceccb6b6faa2 kytos.core.atcp_server:INFO atcp_server:130:  New connection from 172.18.0.4:33746
2026-04-25T11:31:20.024595+00:00 ceccb6b6faa2 kytos.napps.kytos/of_core:INFO main:484:  connection ('172.18.0.4', 33746): negotiated version - 4
2026-04-25T11:31:20.027826+00:00 ceccb6b6faa2 kytos.napps.kytos/of_core:INFO main:499:  Connection ('172.18.0.4', 33746): Hello complete
2026-04-25T11:31:20.030862+00:00 ceccb6b6faa2 kytos.napps.kytos/of_core:INFO main:367:  Connection ('172.18.0.4', 33746): IN OFP, ver: 4, type: 6, xid: 1159978633
2026-04-25T11:31:20.031314+00:00 ceccb6b6faa2 kytos.napps.kytos/of_core:INFO main:379:  unprocessed packet for ('172.18.0.4', 33746) sending_features ConnectionState.SETUP
```

We can see that switch 03 finished the OF version negotiation and Kytos sent the FeaturesRequest message around `11:31:20.0278`. The Features Reply was received at `11:31:20.0308`, but then of_core skipped processing that message at `11:31:20.0313` because the connection state was still on `sending_features` instead of `waiting_features_reply` with this IF statement: https://github.com/kytos-ng/of_core/blob/master/main.py#L373-L375

What probably happened at this case was that OF Features Request and Reply were exchanged sooner than the Kytos Message Bug got a change to actually change the connection state machine (https://github.com/kytos-ng/of_core/blob/1febd118f5894bae8633c7770b10d4ac9a4b9edf/main.py#L541-L544)

Furthermore, it seems the idea of adding that message to the list of unprocessed_packets and then connection.remaining_data is not effective, once the connection will enter into a "quiet" mode were both parts (switch and conntroller) wont exchange messages, and then the remaining data will never be processed again.

Thus, I see two problems here:

1) we should make this connection state transition more robust so that delays on the message bus dont impact the connection setup itself.

2) we should review the approach of processing the remaining_data for cases where no new messages wont trigger the loop but we still have data to be processed.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -11,6 +11,8 @@ Changed
 - Added ``ipaddress`` lib to parser requests for IPv4 and IPv4 on MatchBase,
 allowing validation of IPv6 compressed addresses, IPv4 with /32 masks and
 IPv4 network with host bit set.
+- Adding ``sending_features`` as a valid state for waiting features during
+OF Handshake phase.
 
 [2025.2.0] - 2026-02-02
 ***********************

--- a/main.py
+++ b/main.py
@@ -368,7 +368,10 @@ class Main(KytosNApp):
                 ofp_msg_type_str = message.header.message_type.name.lower()
                 waiting_features_reply = (
                     ofp_msg_type_str == 'ofpt_features_reply'
-                    and connection.protocol.state == 'waiting_features_reply')
+                    and connection.protocol.state in (
+                        'waiting_features_reply', 'sending_features'
+                    )
+                )
 
                 if connection.is_during_setup() and not waiting_features_reply:
                     unprocessed_packets.append(packet)

